cmd/speedtest/speedtest.go

```diff
@@ -0,0 +1,193 @@
+// This will start a speedtest server on 127.0.0.1:8080 and allow only 1 concurrent connection at a time.
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"tailscale.com/client/tailscale"
+	"tailscale.com/net/speedtest"
+
+	"github.com/peterbourgon/ff/v2/ffcli"
+)
+
+// Runs the speedtest command as a commandline program
+func main() {
+	args := os.Args[1:]
+	if err := speedtestCmd.Parse(args); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	err := speedtestCmd.Run(context.Background())
+	if err != nil {
+		if err != flag.ErrHelp {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+	}
+}
+
+// Speedtest command that contains the server and client sub commands.
+var speedtestCmd = &ffcli.Command{
+	Name:       "speedtest",
+	ShortUsage: "speedtest {server|client} ...",
+	ShortHelp:  "Run a speed test",
+	Subcommands: []*ffcli.Command{
+		speedtestServerCmd,
+		speedtestClientCmd,
+	},
+	Exec: func(context.Context, []string) error {
+		return errors.New("subcommand required; run 'tailscale speedtest -h' for details")
+	},
+}
+
+// speedtestServerCmd takes necessary info like the port to
+// listen on and then passes them to the StartServer function in the speedtest package.
+// if the localhost flag is given, the server will use 127.0.0.1, otherwise the server will
+// use the tailscale ip address
+var speedtestServerCmd = &ffcli.Command{
+	Name:       "server",
+	ShortUsage: "speedtest server --host <host> --port <port> [--max-conns <max connections>]",
+	ShortHelp:  "Start a speed test server",
+	Exec:       runServer,
+	FlagSet: (func() *flag.FlagSet {
+		fs := flag.NewFlagSet("server", flag.ExitOnError)
+		fs.IntVar(&serverArgs.port, "port", 0, "port to listen on")
+		fs.BoolVar(&serverArgs.localhost, "localhost", false, "use localhost or tailscale ip")
+		fs.IntVar(&serverArgs.maxConnections, "max-conns", 1, "max number of concurrent connections allowed")
+		return fs
+	})(),
+}
+
+// speedtestClientCmd takes info like the type of test to run, message size, test time, and the host and port
+// of the speedtest server and passes them to the StartClient function in the speedtest package.
+var speedtestClientCmd = &ffcli.Command{
+	Name:       "client",
+	ShortUsage: "speedtest client {--d|--u} --host <host> --port <port> [--inc <increment>] [--seconds <number of seconds>]",
+	ShortHelp:  "Start a speed test client and connect to a speed test server",
+	Exec:       runClient,
+	FlagSet: (func() *flag.FlagSet {
+		fs := flag.NewFlagSet("client", flag.ExitOnError)
+		fs.StringVar(&clientArgs.host, "host", "", "The ip address for the speedtest server being used")
+		fs.StringVar(&clientArgs.port, "port", "", "The port of the speedtest server being used")
+		fs.IntVar(&clientArgs.inc, "inc", 1, "The increment for displaying speedtest info")
+		fs.BoolVar(&clientArgs.download, "d", false, "Include this to run a download test")
+		fs.BoolVar(&clientArgs.upload, "u", false, "Include this to run an upload test")
```

Comment:
As we discussed on video, I don't think you need separate upload and download modes. We'd handle that in the function which invokes the speedtest from tailscaled.